### ISSUE TITLE
modkit call-mods omits all modification information for many RNA004 reads 

### ISSUE THREAD

**Issue by pre-mRNA** [nanoporetech/modkit#202]
**Edit: The issue was caused by my incorrect use of 'filter-threshold' argument in place of 'mod-threshold' argument. Please feel free to delete the issue.** 

Hi, 

Thanks for developing modkit, which is a useful tool for our work. 

I'm using modkit v0.3.1 to call-mods for RNA004 reads aligned to the human genome. 

The command I use is: 
```
modkit call-mods -t 104 --filter-threshold 0.995 $in_bam $out_bam
```

On IGV browser, I notice that many reads that originally had modification predictions lose all their modification information after calling mods. This seems to effect mainly longer RNA004 reads. 

My example input read, 7f1e9aaa-7e59-4948-9296-53f9c4b16948, has modifications calls: 

MM:Z:A+a.,0,2,1,6,0,0,0,2,0,0,2,4,1,0,1,6,9,4,2,0,1,0,9,0,0,8,0,3,1,0,20,4,0,8,4,15,0,2,5,2,2,4,1,0,1,3,0,0,0,0,0,0,3,0,0,0,9,0,0,0,2,8,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0;       ML:B:C,88,249,229,17,12,31,17,251,64,30,91,34,29,55,23,25,32,35,89,56,226,15,19,85,73,61,96,93,97,108,63,85,31,19,18,98,96,46,40,51,32,23,25,48,75,37,61,22,197,141,124,13,85,164,123,27,104,235,111,30,21,13,110,115,156,163,163,172,171,171,173,174,187,180,177,175,179,189,168

But in the mod-called bamfile, these calls are missing: 

 MM:Z:A+a.;      ML:B:C

I attach the example read in BAM format so you can test it.

[7f1e9aaa-7e59-4948-9296-53f9c4b16948.zip](https://github.com/user-attachments/files/15783330/7f1e9aaa-7e59-4948-9296-53f9c4b16948.zip)

Thanks a lot!

**Comment by pre-mRNA** [comment]
There's nothing in the log for the failed read: 

[src/logging.rs::60][2024-06-11 16:52:24][DEBUG] command line: modkit call-mods -t 104 --filter-threshold 0.995 $in_bam --log-filepath $log
[src/adjust.rs::179][2024-06-11 16:52:24][INFO] done, 2 records processed, 0 failed, 0 skipped

The log says that it processed two reads, even though the BAM file I shared only contains a single alignment. 

Dropping the filter threshold to 0.9 seems to fix the issue for this specific read, but the cause for this behaviour isn't clear to me. 

Thanks!

**Comment by ArtRand** [comment]
Hello @pre-mRNA, 

Thank you for reporting this issue. You're getting the results I would expect but there is a bug here. 

If you run `extract` you can look at the number of calls above the threshold:

```bash
# just look at the table
$ modkit extract 7f1e9aaa.bam null --read-calls stdout --filter-threshold 0.995
# if you look at this table, you'll see that all of the m6A calls are filtered out (don't pass the threshold)
$ modkit extract 7f1e9aaa.bam null --read-calls stdout --filter-threshold 0.995 2> /dev/null | awk '(NR==1) || ($12=="a")' | column -t | less -S
```

_However_, the bug is that the MM tag should be changed to the explicit `?` mode. I've attached a build that has the fix, here are the modified base tags that are output. Specifically notice that the MM tag has `?` mode, meaning that bases without modified probabilities have "no information" rather than implied to be canonical.

```text
MM:Z:A+a?,1,0,1,1,0,0,0,0,0,4,0,3,0,1,0,0,0,1,2,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,1,0,2,2,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,2,0,0,1,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,2,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,1,0,0,0,0,1,0,1,0,1,0,0,0,1,2,1,0,0,7,0,0,4,0,0,0,0,0,0,0,0,4,0,1,0,0,0,0,0,0,0;
ML:B:C,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
```
[modkit_dev2e9f375_centos7_x86_64.tar.gz](https://github.com/user-attachments/files/15791643/modkit_dev2e9f375_centos7_x86_64.tar.gz)

**Comment by pre-mRNA** [comment]
Bumping this question

**Comment by ArtRand** [comment]
Hello @pre-mRNA 

> Is the default behaviour of Dorado to omit from the BAM reads that are below the filter threshold?

I don't believe that `dorado` filters any reads by default. However, `modkit pileup` will only use primary mappings and will skip reads where it fails to get base modification information. Skipping a non-primary SAM record is silent (there can be a lot of them), and you can pre-filter by excluding SAM flag `2308`. Records that are skipped due to an error are logged in the debug log which is enabled with the `--log-filepath` option.

**Comment by ArtRand** [comment]
This fix has been released in [v0.3.1](https://github.com/nanoporetech/modkit/releases/tag/v0.3.1).
